src/type_checker/fn_env.rs

```diff
@@ -57,13 +57,19 @@ pub struct TypedFnEnv {
     forbid_forloop_scope: bool,
 
     /// The kind of function we're currently type checking
-    current_fn_kind: Option<FuncOrMethod>,
+    current_fn_kind: FuncOrMethod,
 }
 
 impl TypedFnEnv {
-    /// Creates a new TypeEnv
-    pub fn new() -> Self {
-        Self::default()
+    /// Creates a new TypeEnv with the given function kind
+    pub fn new(fn_kind: &FuncOrMethod) -> Self {
```

Comment:
I am sure there is a more elegant way of doing this. The reason I went with this is that since current_fn_kind is not an option, calling Self::default() calls the default constructor of FuncOrMethod as well, which calls unreachable!:

https://github.com/zksecurity/noname/blob/b17ea203e24552974c499ee17bae3fa85b044247/src/parser/types.rs#L739

How can we do it better?

Reply:
Good catch.
With `FuncOrMethod`, I think it should be always calling ::new to specify the `fn_kind`.

Reply:
I think I didn't fully understand the last part. Do you mean the default() for `FuncOrMethod` should change (return a new object instead of unreachable) or is the current approach okay as is?

Reply:
I meant default() should be removed, if it is not used anymore.